onnxscript/ir/_linked_list.py

```diff
@@ -184,13 +178,13 @@ def _insert_one_after(
             return box
         if box.owning_list is not self:
             raise ValueError(f"Value {box.value!r} is not in the list")
-        # Remove the new value from the list if it is already in a different list
-        if new_value._link_box is not None:
-            new_value._link_box.owning_list.remove(new_value)
+
+        if new_value in self._values_to_boxes:
```

Comment:
The implementation seems to reflect an assumption that this is semantically a "set" as opposed to a "list". While that is true for our intended application, it seems unusual for a doubly-linked-list implementation. Just commenting and still thinking. Is there any difference if we use `id(value)` instead of `value`?

Reply:
I guess this stems from the methods like remove and insert_after ...

Reply:
id can be tricky and may cause unexpected behavior:

```
>>> a = (1,)
>>> b = (1,)
>>> id(a)
2649810036928
>>> id(b)
2649815823504
>>> c = "s"
>>> d = "s"
>>> id(c)
2649810832368
>>> id(d)
2649810832368
```

Here to tuples are equal but they have different ids; two strings are equal and they have the same id. I suppose for our use cases (nodes) they would be the same. Using the value as keys makes the "set" behavior uniform across different types.

Reply:
I renamed the class to "[...]HashList" hoping to reflect this behavior. Didn't use "SetList" because we don't expose methods one would find in a set like `add`. Happy to take a better name.

Reply:
We can callout that the DoubleLinkList requires and assumes all elements it holds to be unique.

If we really want to make the class behave like a general linked list, we can make `_values_to_boxes` point to a list of boxes.

Reply:
> We can callout that the DoubleLinkList requires and assumes all elements it holds to be unique.

Right

> If we really want to make the class behave like a general linked list, we can make `_values_to_boxes` point to a list of boxes.

The map is for O(1) retrieval of the boxes from values. I wonder if there are other ways?

I also consider uniqueness to be a feature of this data structure, since it is something we want to maintain in the nodes list.

Reply:
> id can be tricky and may cause unexpected behavior:

My interpretation of `id` is that it is similar to a physical-address (pointer) that remains unique for the object throughout its lifetime. It is perfectly consistent with your examples ... using `id` is acceptable when we expect objects to be unique, and value-based-equality/hash etc. is preferable when we expect we may create distinct objects that will compare equal.

What does the default-implementation of hash/equal do for Nodes? Is it the same as equality of id(...) or is it anything different?

Reply:
I guess the important point is to avoid any complex has/equality on Nodes when using this underlying dictionary.

Reply:
Right. We need to make sure we don't define `__hash__` or equality for Nodes. The default hash of an object is some function of the id(). I am now thinking we should use id() to make this robust. Since we only care about classes we define (not really native python types), id() seems more beneficial.